{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Collecting pytorch-lightning-bolts==0.3.0\n",
      "  Obtaining dependency information for pytorch-lightning-bolts==0.3.0 from https://files.pythonhosted.org/packages/6d/c0/f600b26020bc74e24d3e12163a3a4730b269dc7095643750f4665707c526/pytorch_lightning_bolts-0.3.0-py3-none-any.whl.metadata\n",
      "  Downloading pytorch_lightning_bolts-0.3.0-py3-none-any.whl.metadata (12 kB)\n",
      "Requirement already satisfied: torch>=1.6 in d:\\software\\anaconda3\\lib\\site-packages (from pytorch-lightning-bolts==0.3.0) (2.1.2+cu121)\n",
      "Requirement already satisfied: pytorch-lightning>=1.1.1 in d:\\software\\anaconda3\\lib\\site-packages (from pytorch-lightning-bolts==0.3.0) (2.2.2)\n",
      "Requirement already satisfied: numpy>=1.17.2 in d:\\software\\anaconda3\\lib\\site-packages (from pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (1.24.3)\n",
      "Requirement already satisfied: tqdm>=4.57.0 in d:\\software\\anaconda3\\lib\\site-packages (from pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (4.65.0)\n",
      "Requirement already satisfied: PyYAML>=5.4 in d:\\software\\anaconda3\\lib\\site-packages (from pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (6.0)\n",
      "Requirement already satisfied: fsspec[http]>=2022.5.0 in d:\\software\\anaconda3\\lib\\site-packages (from pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (2023.12.2)\n",
      "Requirement already satisfied: torchmetrics>=0.7.0 in d:\\software\\anaconda3\\lib\\site-packages (from pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (1.3.2)\n",
      "Requirement already satisfied: packaging>=20.0 in d:\\software\\anaconda3\\lib\\site-packages (from pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (23.1)\n",
      "Requirement already satisfied: typing-extensions>=4.4.0 in d:\\software\\anaconda3\\lib\\site-packages (from pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (4.7.1)\n",
      "Requirement already satisfied: lightning-utilities>=0.8.0 in d:\\software\\anaconda3\\lib\\site-packages (from pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (0.11.2)\n",
      "Requirement already satisfied: filelock in d:\\software\\anaconda3\\lib\\site-packages (from torch>=1.6->pytorch-lightning-bolts==0.3.0) (3.9.0)\n",
      "Requirement already satisfied: sympy in d:\\software\\anaconda3\\lib\\site-packages (from torch>=1.6->pytorch-lightning-bolts==0.3.0) (1.11.1)\n",
      "Requirement already satisfied: networkx in d:\\software\\anaconda3\\lib\\site-packages (from torch>=1.6->pytorch-lightning-bolts==0.3.0) (3.1)\n",
      "Requirement already satisfied: jinja2 in d:\\software\\anaconda3\\lib\\site-packages (from torch>=1.6->pytorch-lightning-bolts==0.3.0) (3.1.2)\n",
      "Requirement already satisfied: requests in c:\\users\\trial\\appdata\\roaming\\python\\python311\\site-packages (from fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (2.31.0)\n",
      "Requirement already satisfied: aiohttp!=4.0.0a0,!=4.0.0a1 in d:\\software\\anaconda3\\lib\\site-packages (from fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (3.8.5)\n",
      "Requirement already satisfied: setuptools in d:\\software\\anaconda3\\lib\\site-packages (from lightning-utilities>=0.8.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (60.2.0)\n",
      "Requirement already satisfied: colorama in d:\\software\\anaconda3\\lib\\site-packages (from tqdm>=4.57.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (0.4.6)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in d:\\software\\anaconda3\\lib\\site-packages (from jinja2->torch>=1.6->pytorch-lightning-bolts==0.3.0) (2.1.1)\n",
      "Requirement already satisfied: mpmath>=0.19 in d:\\software\\anaconda3\\lib\\site-packages (from sympy->torch>=1.6->pytorch-lightning-bolts==0.3.0) (1.3.0)\n",
      "Requirement already satisfied: attrs>=17.3.0 in d:\\software\\anaconda3\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (22.1.0)\n",
      "Requirement already satisfied: charset-normalizer<4.0,>=2.0 in d:\\software\\anaconda3\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (2.0.4)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in d:\\software\\anaconda3\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (6.0.2)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in d:\\software\\anaconda3\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (4.0.2)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in d:\\software\\anaconda3\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (1.8.1)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in d:\\software\\anaconda3\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (1.3.3)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in d:\\software\\anaconda3\\lib\\site-packages (from aiohttp!=4.0.0a0,!=4.0.0a1->fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (1.2.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in d:\\software\\anaconda3\\lib\\site-packages (from requests->fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (3.4)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in d:\\software\\anaconda3\\lib\\site-packages (from requests->fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in d:\\software\\anaconda3\\lib\\site-packages (from requests->fsspec[http]>=2022.5.0->pytorch-lightning>=1.1.1->pytorch-lightning-bolts==0.3.0) (2024.2.2)\n",
      "Downloading pytorch_lightning_bolts-0.3.0-py3-none-any.whl (247 kB)\n",
      "   ---------------------------------------- 0.0/247.3 kB ? eta -:--:--\n",
      "   ---------------------------------------- 0.0/247.3 kB ? eta -:--:--\n",
      "   - -------------------------------------- 10.2/247.3 kB ? eta -:--:--\n",
      "   - -------------------------------------- 10.2/247.3 kB ? eta -:--:--\n",
      "   - -------------------------------------- 10.2/247.3 kB ? eta -:--:--\n",
      "   --------- ----------------------------- 61.4/247.3 kB 365.7 kB/s eta 0:00:01\n",
      "   ----------------- -------------------- 112.6/247.3 kB 504.4 kB/s eta 0:00:01\n",
      "   ------------------------------- ------ 204.8/247.3 kB 778.2 kB/s eta 0:00:01\n",
      "   -------------------------------------- 247.3/247.3 kB 894.5 kB/s eta 0:00:00\n",
      "Installing collected packages: pytorch-lightning-bolts\n",
      "  Attempting uninstall: pytorch-lightning-bolts\n",
      "    Found existing installation: pytorch-lightning-bolts 0.2.5\n",
      "    Uninstalling pytorch-lightning-bolts-0.2.5:\n",
      "      Successfully uninstalled pytorch-lightning-bolts-0.2.5\n",
      "Successfully installed pytorch-lightning-bolts-0.3.0\n"
     ]
    }
   ],
   "source": [
    "! pip install pytorch-lightning-bolts==0.3.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pytorch_lightning as pl\n",
    "from torch import nn\n",
    "from torch.nn import functional as F\n",
    "from pl_bolts.models.autoencoders.components import (\n",
    "    resnet18_decoder,\n",
    "    resnet18_encoder,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "class VAE(pl.LightningModule):\n",
    "    def __init__(self, enc_out_dim=512, latent_dim=256, input_height=32):\n",
    "        super().__init__()\n",
    "\n",
    "        self.save_hyperparameters()\n",
    "\n",
    "        # encoder, decoder\n",
    "        self.encoder = resnet18_encoder(False, False)\n",
    "        self.decoder = resnet18_decoder(\n",
    "            latent_dim=latent_dim, \n",
    "            input_height=input_height, \n",
    "            first_conv=False, \n",
    "            maxpool1=False\n",
    "        )\n",
    "\n",
    "        # distribution parameters\n",
    "        self.fc_mu = nn.Linear(enc_out_dim, latent_dim)\n",
    "        self.fc_var = nn.Linear(enc_out_dim, latent_dim)\n",
    "\n",
    "        # for the gaussian likelihood\n",
    "        self.log_scale = nn.Parameter(torch.Tensor([0.0]))\n",
    "\n",
    "    def configure_optimizers(self):\n",
    "        return torch.optim.Adam(self.parameters(), lr=1e-4)\n",
    "\n",
    "    def gaussian_likelihood(self, mean, logscale, sample):\n",
    "        scale = torch.exp(logscale)\n",
    "        dist = torch.distributions.Normal(mean, scale)\n",
    "        log_pxz = dist.log_prob(sample)\n",
    "        return log_pxz.sum(dim=(1, 2, 3))\n",
    "\n",
    "    def kl_divergence(self, z, mu, std):\n",
    "        # --------------------------\n",
    "        # Monte carlo KL divergence\n",
    "        # --------------------------\n",
    "        # 1. define the first two probabilities (in this case Normal for both)\n",
    "        p = torch.distributions.Normal(torch.zeros_like(mu), torch.ones_like(std))\n",
    "        q = torch.distributions.Normal(mu, std)\n",
    "\n",
    "        # 2. get the probabilities from the equation\n",
    "        log_qzx = q.log_prob(z)\n",
    "        log_pz = p.log_prob(z)\n",
    "\n",
    "        # kl\n",
    "        kl = (log_qzx - log_pz)\n",
    "        kl = kl.sum(-1)\n",
    "        return kl\n",
    "\n",
    "    def training_step(self, batch, batch_idx):\n",
    "        x, _ = batch\n",
    "\n",
    "        # encode x to get the mu and variance parameters\n",
    "        x_encoded = self.encoder(x)\n",
    "        mu, log_var = self.fc_mu(x_encoded), self.fc_var(x_encoded)\n",
    "\n",
    "        # sample z from q\n",
    "        std = torch.exp(log_var / 2)\n",
    "        q = torch.distributions.Normal(mu, std)\n",
    "        z = q.rsample()\n",
    "\n",
    "        # decoded \n",
    "        x_hat = vae.decoder(z)\n",
    "\n",
    "        # reconstruction loss\n",
    "        recon_loss = self.gaussian_likelihood(x_hat, self.log_scale, x)\n",
    "\n",
    "        # kl\n",
    "        kl = self.kl_divergence(z, mu, std)\n",
    "\n",
    "        # elbo\n",
    "        elbo = (kl - recon_loss)\n",
    "        elbo = elbo.mean()\n",
    "\n",
    "        self.log_dict({\n",
    "            'elbo': elbo,\n",
    "            'kl': kl.mean(),\n",
    "            'recon_loss': recon_loss.mean(), \n",
    "            'reconstruction': recon_loss.mean(),\n",
    "            'kl': kl.mean(),\n",
    "        })\n",
    "\n",
    "        return elbo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2.2.2'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pl.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unexpected exception formatting exception. Falling back to standard exception\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3526, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\Trial\\AppData\\Local\\Temp\\ipykernel_20104\\1802927316.py\", line 1, in <module>\n",
      "    from pl_bolts.datamodules import CIFAR10DataModule\n",
      "  File \"C:\\Users\\Trial\\AppData\\Roaming\\Python\\Python311\\site-packages\\pl_bolts\\__init__.py\", line 47, in <module>\n",
      "    from pl_bolts import callbacks, datamodules, datasets, losses, metrics, models, optimizers, transforms, utils\n",
      "  File \"C:\\Users\\Trial\\AppData\\Roaming\\Python\\Python311\\site-packages\\pl_bolts\\callbacks\\__init__.py\", line 5, in <module>\n",
      "    from pl_bolts.callbacks.data_monitor import ModuleDataMonitor, TrainingDataMonitor  # noqa: F401\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"C:\\Users\\Trial\\AppData\\Roaming\\Python\\Python311\\site-packages\\pl_bolts\\callbacks\\data_monitor.py\", line 6, in <module>\n",
      "    from pytorch_lightning.loggers import LightningLoggerBase, TensorBoardLogger, WandbLogger\n",
      "ImportError: cannot import name 'LightningLoggerBase' from 'pytorch_lightning.loggers' (d:\\Software\\Anaconda3\\Lib\\site-packages\\pytorch_lightning\\loggers\\__init__.py)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 2120, in showtraceback\n",
      "    stb = self.InteractiveTB.structured_traceback(\n",
      "          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\IPython\\core\\ultratb.py\", line 1435, in structured_traceback\n",
      "    return FormattedTB.structured_traceback(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\IPython\\core\\ultratb.py\", line 1326, in structured_traceback\n",
      "    return VerboseTB.structured_traceback(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\IPython\\core\\ultratb.py\", line 1173, in structured_traceback\n",
      "    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n",
      "                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\IPython\\core\\ultratb.py\", line 1088, in format_exception_as_a_whole\n",
      "    frames.append(self.format_record(record))\n",
      "                  ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\IPython\\core\\ultratb.py\", line 970, in format_record\n",
      "    frame_info.lines, Colors, self.has_colors, lvals\n",
      "    ^^^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\IPython\\core\\ultratb.py\", line 792, in lines\n",
      "    return self._sd.lines\n",
      "           ^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\stack_data\\utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "                                               ^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\stack_data\\core.py\", line 698, in lines\n",
      "    pieces = self.included_pieces\n",
      "             ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\stack_data\\utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "                                               ^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\stack_data\\core.py\", line 649, in included_pieces\n",
      "    pos = scope_pieces.index(self.executing_piece)\n",
      "                             ^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\stack_data\\utils.py\", line 145, in cached_property_wrapper\n",
      "    value = obj.__dict__[self.func.__name__] = self.func(obj)\n",
      "                                               ^^^^^^^^^^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\stack_data\\core.py\", line 628, in executing_piece\n",
      "    return only(\n",
      "           ^^^^^\n",
      "  File \"d:\\Software\\Anaconda3\\Lib\\site-packages\\executing\\executing.py\", line 164, in only\n",
      "    raise NotOneValueFound('Expected one value, found 0')\n",
      "executing.executing.NotOneValueFound: Expected one value, found 0\n"
     ]
    }
   ],
   "source": [
    "from pl_bolts.datamodules import CIFAR10DataModule\n",
    "\n",
    "datamodule = CIFAR10DataModule('.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pl_bolts.datamodules import CIFAR10DataModule\n",
    "\n",
    "# train_transforms = torchvision.transforms.Compose(\n",
    "#     [\n",
    "#         torchvision.transforms.RandomCrop(32, padding=4),\n",
    "#         torchvision.transforms.RandomHorizontalFlip(),\n",
    "#         torchvision.transforms.ToTensor(),\n",
    "#         cifar10_normalization(),\n",
    "#     ]\n",
    "# )\n",
    "\n",
    "# test_transforms = torchvision.transforms.Compose(\n",
    "#     [\n",
    "#         torchvision.transforms.ToTensor(),\n",
    "#         cifar10_normalization(),\n",
    "#     ]\n",
    "# )\n",
    "\n",
    "# cifar10_dm = CIFAR10DataModule(\n",
    "#     data_dir='.',\n",
    "#     batch_size=16,\n",
    "#     num_workers=2,\n",
    "#     train_transforms=train_transforms,\n",
    "#     test_transforms=test_transforms,\n",
    "#     val_transforms=test_transforms,\n",
    "# )\n",
    "\n",
    "# datamodule = CIFAR10DataModule('.')\n",
    "\n",
    "# traindl = DataLoader(datamodule)\n",
    "\n",
    "# trainset = torchvision.datasets.CIFAR10(root='./data', train=True,\n",
    "#                                         download=True)\n",
    "# sampler = torch.utils.data.distributed.DistributedSampler(\n",
    "#     trainset,\n",
    "#     num_replicas=xm.xrt_world_size(),\n",
    "#     shuffle=True)\n",
    "\n",
    "# trainloader = torch.utils.data.DataLoader(trainset, batch_size=16,\n",
    "#                                           shuffle=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
